designs/component-config.md

```diff
@@ -0,0 +1,318 @@
+type ControllerConfigurationHealth struct {
+	HealthProbeBindAddress string `json:"healthProbeBindAddress,omitempty"`
+
+	ReadinessEndpointName string `json:"readinessEndpointName,omitempty"`
+	LivenessEndpointName  string `json:"livenessEndpointName,omitempty"`
+}
+```
+
+
+
+#### Default ComponentConfig Type
+
+To enable `controller-runtime` to have a default `ComponentConfig` struct which can be used instead of requiring each controller or extension to build it's own `ComponentConfig` type, we can create a `DefaultControllerConfiguration` type which can exist in `pkg/api/config/v1alpha1/types.go`. This will allow the controller authors to use this before needing to implement their own type with additional configs.
+
+```golang
+// pkg/api/config/v1alpha1/types.go
+package v1alpha1
+
+import (
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	configv1alpha1 "sigs.k8s.io/controller-runtime/pkg/apis/config/v1alpha1"
+)
+
+// DefaultControllerConfiguration is the Schema for the DefaultControllerConfigurations API
+type DefaultControllerConfiguration struct {
+	metav1.TypeMeta `json:",inline"`
+
+	Spec   configv1alpha1.ControllerConfiguration   `json:"spec,omitempty"`
+}
+```
+
+This would allow a controller author to use this struct with any config that supports the json/yaml structure. For example a controller author could define their `Kind` as `FoobarControllerConfiguration` and have it defined as the following.
+
+```yaml
+# config.yaml
+apiVersion: config.somedomain.io/v1alpha1
+kind: FoobarControllerConfiguration
+spec:
+  port: 9443
+  metricsBindAddress: ":8080"
+  leaderElection:
+    leaderElect: false
+```
+
+Given the following config and `DefaultControllerConfiguration` we'd be able to initialize the controller using the following.
+
+
+```golang
+mgr, err := ctrl.NewManagerFromComponentConfig(ctrl.GetConfigOrDie(), scheme, configname, &defaultv1alpha1.DefaultControllerConfiguration{})
+if err != nil {
+	// ...
+}
+```
+
+The above example uses `configname` which is the name of the file to load the configuration from and uses `scheme` to get the specific serializer, eg `serializer.NewCodecFactory(scheme)`. This will allow the configuration to be unmarshalled into the `runtime.Object` type and passed into the 
+`ctrl.NewManagerFromComponentConfig()` as a `ManagerConfiguration` interface.
+
+#### Using Flags w/ ComponentConfig
+
+Since this design still requires setting up the initial `ComponentConfig` type and passing in a pointer to `ctrl.NewFromComponentConfig()` if you want to allow for the use of flags, your controller can use any of the different flagging interfaces. eg [`flag`](https://golang.org/pkg/flag/), [`pflag`](https://godoc.org/github.com/spf13/pflag), [`flagnum`](https://godoc.org/github.com/luci/luci-go/common/flag/flagenum) and set values on the `ComponentConfig` type prior to passing the pointer into the `ctrl.NewFromComponentConfig()`, example below.
```

Comment:
Just FYI for backwards compatibility flags usually need to take precedence over config (unless you can force users to use exclusively flags or config).

Reply:
Good callout, since we're decoding in the same object that the flags could have already been set on there isn't a way I'm aware of to not override existing values while decoding. Have any suggestions?

Reply:
save off the flags and apply them after the decode? (merge the options structs together more or less)?

Reply:
Yea, that probably makes the most sense. I personally don't think this needs to be documented in the design doc, do you? 

Feeling okay with the rest of this @shawn-hurley ?

Reply:
Instead of `ctrl.NewFromComponentConfig()` that returns `Manager`, would it be reasonable to have `ctrl.NewOptionsFromComponentConfig()`  that returns `manager.Options`. And then let callers call `manager.New()` themselves so that they have more flexibility in updating `manager.Options` after it has been populated?

Reply:
Totally a possibility. I'd originally started with that path and ended up combining the funcs to make the new API quicker to implement. 

@shawn-hurley @mtaufen @DirectXMan12 @stealthybox  Do you have a preference? The options would be:

1.  Use a single function `ctrl.NewFromComponentConfig()` which takes the `ComponentConfig` type and returns a `ctrl.Manager`.

### or

2. Separate the funcs and make it `ctrl.NewOptionsFromComponentConfig()` which returns `manager.Options` that implementation wise would be passed into `manager.New()`.

Reply:
Another thought would be for the ComponentConfig-populating function to accept a pointer to a `manager.Options` struct so that something like this would be possible:

```go
func main() {
   opts := manager.Options{}
   _ = flag.StringVar(&opts.Namespace, "namespace", "", "Namespace to watch.")

   ctrl.UpdateOptionsFromComponentConfig(&opts)

   // Parse flags after populating from ComponentConfig so that flags
   // take precedence
   _ = flag.Parse()
   _ = manager.New(ctrl.GetConfigOrDie(), opts)
}

Reply:
Yeah, I think both are fair flows. I think it might be easier to guarantee that flags take precedence by decoupling the parsing and new manager initialization logic. This also opens the door later to add the `ctrl.NewManagerFromComponentConfig()`. Where as it would be hard to walk backwards from having both manager init and populating the `Options{}` struct.

Does anyone have strong opinions against this? @stealthybox @johnsonj @joelanford @djzager @shawn-hurley @mtaufen